core/tf_psa_crypto_version.c

```diff
@@ -0,0 +1,31 @@
+/*
+ *  Version information
+ *
+ *  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0 OR GPL-2.0-or-later
+ */
+
+#include "common.h"
+
+#if defined(TF_PSA_CRYPTO_VERSION)
+#include "tf-psa-crypto/version.h"
+#include <string.h>
+
+unsigned int tf_psa_crypto_version_get_number(void)
+{
+    return TF_PSA_CRYPTO_VERSION_NUMBER;
+}
+
+void tf_psa_crypto_version_get_string(char *string)
+{
+    memcpy(string, TF_PSA_CRYPTO_VERSION_STRING,
+           sizeof(TF_PSA_CRYPTO_VERSION_STRING));
+}
+
+void tf_psa_crypto_version_get_string_full(char *string)
+{
+    memcpy(string, TF_PSA_CRYPTO_VERSION_STRING_FULL,
+           sizeof(TF_PSA_CRYPTO_VERSION_STRING_FULL));
+}
```

Comment:
I know that this is the interface we use in Mbed TLS, but is this really the best way to go about it?
Our users have no way of knowing how large the buffer needs to be to contain our version strings, aside from making educated guesses.

String literals have a static storage duration in both [C](https://en.cppreference.com/w/c/language/string_literal.html) and [C++](https://en.cppreference.com/w/cpp/language/string_literal.html), so returning a pointer to a string literal should be fine when compiled with any compiler of either language.

```suggestion
char *tf_psa_crypto_version_get_string(void)
{
    return TF_PSA_CRYPTO_VERSION_STRING;
}

char *tf_psa_crypto_version_get_string_full(void)
{
    return TF_PSA_CRYPTO_VERSION_STRING_FULL;
}
```

@gilles-peskine-arm Thoughts?

Reply:
I agree. Having a fixed-size string limits us. We can't rename the product. We only have room for a certain number of digits, no way to add a suffix like `beta` and we should be careful with numbers. I'm all for just returning a pointer to a global constant.

Reply:
I've even [added it to the 4.0 shoulds](https://github.com/Mbed-TLS/mbedtls/issues/10308). It's not very important, but it's very small.

Reply:
> for the sake of consistency

I don't see any need for consistency here. If there was a `get_version_of_crypto_library()` function that already existed in Mbed TLS then TF-PSA-Crypto should maintain backward compatibility with that, to allow applications to work with either product. But there wasn't, there was only “get version of Mbed TLS” and TF-PSA-Crypto doesn't offer a version of Mbed TLS. So “get version of the crypto library” is a new thing anyway.

Reply:
Yes, I understood that, sorry for being unclear. My point is that the only benefit I see for consistency is to simplify things for callers who want to get the crypto version in code that needs to work with both 3.6 and 1.0, and having similar prototypes with just different function names and buffer sizes doesn't really simplify that. If anything, callers might just change the function name and forget to make the buffer large enough for `"TF-PSA-Crypto 1.23.4\000"`.

Reply:
Since David is on Leave now, I would be willing to take this one over, and raise an equivalent PR updating to the version functions in 4.0.

Reply:
David is still there today.

Reply:
Ah, ok. My bad.

Reply:
That's a really good point, thanks!